src/utils/exec.c

```diff
@@ -733,13 +734,77 @@ gboolean bd_utils_init_prog_reporting (BDUtilsProgFunc new_prog_func, GError **e
     return TRUE;
 }
 
+/**
+ * bd_utils_init_prog_reporting_thread:
+ * @new_prog_func: (allow-none) (scope notified): progress reporting function to
+ *                                                use on current thread or %NULL
+ *                                                to reset to default or global
+ * @error: (out): place to store error (if any)
+ *
+ * Returns: whether progress reporting was successfully initialized or not
+ */
+gboolean bd_utils_init_prog_reporting_thread (BDUtilsProgFunc new_prog_func, GError **error __attribute__((unused))) {
+    /* XXX: the error attribute will likely be used in the future when this
+       function gets more complicated */
+
+    thread_prog_func = new_prog_func;
+
+    return TRUE;
+}
+
+static void thread_progress_deinitialized (guint64 task_id __attribute__((unused)), BDUtilsProgStatus status __attribute__((unused)), guint8 completion __attribute__((unused)), gchar *msg __attribute__((unused))) {
+    /* This function serves as a special value for the progress reporting
+     * function to detect that nothing is done here. If clients use their own
+     * empty function then bd_utils_prog_reporting_initialized will return TRUE
+     * but with this function here it returns FALSE.
+     */
+}
```

Comment:
Can't we just use ``NULL`` instead of this special function?

Reply:
`NULL` is the default and means that no local reporting function is used and thus the global one is called.

Reply:
But local deinitialization is needed to avoid the parsing overhead. Reversing the process of setting up a reporting function is similar in global and local state: passing `NULL` to `bd_utils_init_prog_reporting_thread` as it is done with `bd_utils_init_prog_reporting`.

Reply:
`NULL` for `bd_utils_init_prog_reporting_thread` resets the behavior to the global initialization which might or might not have been set up.

Reply:
Which is exactly what ``thread_initialize()`` followed by ``thread_deinitialize()`` should do I think.

Reply:
``thread_deinitialize()`` should guarantee that **thread** reporting is deinitialized/not used. So I think it should just set the thread-local reporting function to ``NULL``. So the same as ``thread_initialize(NULL)``, but it provides some level of abstraction and can do different things in the future.

Reply:
Ok, good, was not sure what naming is best. Then what name should the function have that deinitializes reporting for a thread? How about `bd_utils_disable_any_prog_reporting_thread`?

Reply:
Well, that's what ``mute()`` should do, ideally also avoiding the parsing overhead, but we can probably live with it, if necessary.

Reply:
Ah, then I misunderstood you in the beginning with the 'convenience function', sorry… Ok, empty callbacks can be set up by the client but a special value which disables it not, so only this will get a function here called `mute`. I'll update it ;)

Reply:
Do you plan to do some more changes here? Because right now ``thread_deinitialize()`` is not a reverse process to ``thread_initialize()``, it's the same as ``mute()``, right?

Reply:
Hi, sorry, that function was just a leftover and not used anymore. I think this is it now.